ddtrace/contrib/botocore/services/bedrock.py

```diff
@@ -0,0 +1,297 @@
+import json
+import sys
+from typing import Any
+from typing import Dict
+from typing import List
+
+from ddtrace import Span
+from ddtrace.internal.llmobs.integrations import BedrockIntegration
+from ddtrace.vendor import wrapt
+
+from ....internal.schema import schematize_service_name
+
+
+_AI21 = "ai21"
+_AMAZON = "amazon"
+_ANTHROPIC = "anthropic"
+_COHERE = "cohere"
+_META = "meta"
+_STABILITY = "stability"
+
+
+class TracedBotocoreStreamingBody(wrapt.ObjectProxy):
+    #  Currently the corresponding span finishes only if
+    #  1) the user fully consumes the stream body
+    #  2) error during reading
+
+    def __init__(self, wrapped, span, integration):
+        super().__init__(wrapped)
+        self._datadog_span = span
+        self._datadog_integration = integration
+        self._body = []
+
+    def read(self, amt=None):
+        try:
+            body = self.__wrapped__.read(amt=amt)
+            self._body.append(json.loads(body))
+            if self.__wrapped__.tell() == int(self.__wrapped__._content_length):
+                formatted_response = _extract_response(self._datadog_span, self._body[0])
+                self._process_response(formatted_response)
+                self._datadog_span.finish()
+            return body
+        except Exception:
+            self._datadog_span.set_exc_info(*sys.exc_info())
+            self._datadog_span.finish()
+            raise
+
+    def readlines(self):
+        try:
+            lines = self.__wrapped__.readlines()
+            for line in lines:
+                self._body.append(json.loads(line))
+            formatted_response = _extract_response(self._datadog_span, self._body[0])
+            self._process_response(formatted_response)
+            self._datadog_span.finish()
```

Comment:
We should move the span finish to a "finally" block

Reply:
This won't work, since `read(n_bytes)` can read only up to `n_bytes` and not fully consume the stream. We don't want to finish the span always when this method is called, only when the stream is fully read or on error.